Sources/Basics/HTTPClient/URLSessionHTTPClient.swift

```diff
@@ -337,6 +353,270 @@ private class DownloadTaskManager: NSObject, URLSessionDownloadDelegate {
+                    let dataToWrite: LegacyHTTPClientRequest.UploadData?
+
+                    switch self.dataToWrite {
+                    case .none:
+                        switch try self.streamProvider() {
+                        case .chunk(let data):
+                            dataToWrite = .chunk(data)
+                        case .stream(let stream):
+                            dataToWrite = .stream(stream)
+                        case .none:
+                            dataToWrite = .none
+                        }
+                    case .chunk(let data):
+                        dataToWrite = .chunk(data)
+                    case .stream(let stream):
+                        dataToWrite = .stream(stream)
+                    }
+
+                    switch dataToWrite {
+                    case .chunk(let data):
+                        let totalBytes = data.count
+                        let bytesWritten = try data
+                            .withUnsafeBytes { (unsafeRawBufferPointer: UnsafeRawBufferPointer) in
+                                let unsafeBufferPointer = unsafeRawBufferPointer.bindMemory(to: UInt8.self)
+                                guard let unsafePointer = unsafeBufferPointer.baseAddress else {
+                                    throw HTTPClientError
+                                        .uploadError("Invalid upload state, cannot read upload buffer pointer")
+                                }
+                                return self.outputStream.write(
+                                    unsafePointer,
+                                    maxLength: Swift.min(totalBytes, Self.bufferSize)
+                                )
+                            }
+                        if bytesWritten < totalBytes {
+                            self.dataToWrite = .chunk(data.subdata(in: bytesWritten ..< totalBytes))
+                        } else if bytesWritten == totalBytes {
+                            // done with chunk
+                            self.dataToWrite = .none
+                        } else {
+                            throw HTTPClientError
+                                .uploadError("Invalid upload state: sent too many bytes from data chunk")
+                        }
+                    case .stream(let inputStream):
+                        inputStream.open()
+                        if inputStream.hasBytesAvailable {
+                            var buffer = [UInt8](repeating: 0, count: Self.bufferSize)
+                            let bytesRead = inputStream.read(&buffer, maxLength: buffer.count)
+                            if bytesRead == 0 {
+                                // done with stream
+                                self.dataToWrite = .none
+                                inputStream.close()
+                            } else {
+                                let bytesWritten = self.outputStream.write(buffer, maxLength: bytesRead)
```

Comment:
what happens with the bytes that didn't get written? (i.e. `bytesWritten < bytesRead`)?

Reply:
great catch. is there a way to move the reader index back? otherwise I can buffer them

Reply:
No, you can't unread. You'll need to buffer them and you'll need to stop reading until you're done with the previous writes.